File: client/src/main/scala/stasis/client/encryption/secrets/UserHashedAuthenticationPassword.scala
package stasis.client.encryption.secrets

import akka.util.ByteString
import stasis.shared.model.users.User
import stasis.shared.secrets.DerivedPasswords

import java.util.concurrent.atomic.AtomicBoolean

// doc - sent to auth provider
final case class UserHashedAuthenticationPassword(
  user: User.Id,
  private val hashedPassword: ByteString
) extends Secret {
  private val extracted: AtomicBoolean = new AtomicBoolean(false)

  @SuppressWarnings(Array("org.wartremover.warts.Throw"))
  def extract(): String = {
    val alreadyExtracted = extracted.getAndSet(true)

    if (alreadyExtracted) {
      throw new IllegalStateException("Password already extracted")
    } else {
      DerivedPasswords.encode(hashedPassword)
    }
  }
}

object UserHashedAuthenticationPassword {
  def apply(user: User.Id, hashedPassword: ByteString): UserHashedAuthenticationPassword =
    new UserHashedAuthenticationPassword(user, hashedPassword)
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/encryption/secrets/UserHashedAuthenticationPassword.kt
package stasis.client_android.lib.encryption.secrets

import okio.ByteString
import stasis.client_android.lib.model.server.users.UserId
import java.util.Base64
import java.util.concurrent.atomic.AtomicBoolean

data class UserHashedAuthenticationPassword(
    val user: UserId,
    private val hashedPassword: ByteString
) : Secret() {
    private val extracted: AtomicBoolean = AtomicBoolean(false)

    fun extract(): String {
        val alreadyExtracted = extracted.getAndSet(true)

        return if (alreadyExtracted) {
            throw IllegalStateException("Password already extracted")
        } else {
            Base64.getUrlEncoder().withoutPadding().encodeToString(hashedPassword.toByteArray())
        }
    }
}
